include/device.h

```diff
@@ -577,16 +663,53 @@ static inline int device_pm_put_sync(struct device *dev) { return -ENOTSUP; }
  * @}
  */
 
+/** Synthesize the name of the object that holds device ordinal and
+ * dependency data.  If the object doesn't come from a devicetree
+ * node, use dev_name */
+#define Z_DEVICE_ORD_NAME(node_id,dev_name) _CONCAT(__device_ords, COND_CODE_1(DT_NODE_EXISTS(node_id),(node_id),(dev_name)))
+
+#define Z_DEVICE_EXTRA_ORDS(...) FOR_EACH_NONEMPTY_TERM(IDENTITY, (,), __VA_ARGS__)
+
+/* Construct the dependency ordinal array for the device.
+ * Still need a way to inject handles that don't come from devicetree. */
```

Comment:
I think that this means you want to pass some extra ordinals via `...`, but you need two lists instead of just one, and `...` expands to just one (`__VA_ARGS__`).

If I got that right, is there a way to pass these in from the calling macros via parenthesized lists, and use `__DEBRACKET` to pull the parens off?

The DEBRACKETED_HANDLES macro defined here is what I mean, along with examples of how to use it in various situations: https://github.com/mbolivar-nordic/zephyr/commit/537a0aded7a10d6c6cf469c17c6fd31bc90cef71

If you can define Z_DEVICE_DEFINE_DT_PRE so that it takes a fixed number of parenthesized lists and make the callers play along, it should work here also. Something like this:

```
#define Z_DEVICE_DEFINE_DT_PRE(node_id, dev_name, extra_ords, other_ord_set)		\
...
			DT_REQUIRES_DEP_ORDS(node_id)			\
		),( 							\
			__DEBRACKET other_ord_set			\
		))							\
			Z_DEVICE_EXTRA_ORDS(__DEBRACKET extra_ords)	\
			DEVICE_HANDLE_SEP,				\
	COND_CODE_1(DT_NODE_EXISTS(node_id), (				\
...
```

If I didn't get you, sorry for the noise.

Reply:
it's a single value, but as currently designed i needs to be an argument to `SYS_DEVICE_DEFINE()` that has no counterpart in any other `DEVICE_DEFINE()`, and it needs to be passed through to `Z_DEVICE_DEFINE_DT_PRE` to replace the `NULL` at the "inject from caller" comment, if it's provided.

Changing the trailing args to be bracketed could be a solution; thanks.

In the complex but optimized solution it probably wouldn't even appear in the code or the macros, but instead would be assigned and recorded in the post-process tooling.